tests/performance/CatalogClientTest.cpp

```diff
@@ -0,0 +1,213 @@
+    return olp::http::NetworkProxySettings()
+        .WithHostname("http://localhost:3000")
+        .WithUsername("test_user")
+        .WithPassword("test_password")
+        .WithType(olp::http::NetworkProxySettings::Type::HTTP);
+  }
+
+  olp::client::OlpClientSettings CreateCatalogClientSettings() {
+    auto parameter = GetParam();
+
+    std::shared_ptr<olp::thread::TaskScheduler> task_scheduler = nullptr;
+    if (parameter.task_scheduler_capacity != 0) {
+      task_scheduler =
+          olp::client::OlpClientSettingsFactory::CreateDefaultTaskScheduler(
+              parameter.task_scheduler_capacity);
+    }
+
+    olp::authentication::Settings settings;
+    settings.task_scheduler = task_scheduler;
+    settings.network_request_handler = s_network;
+
+    olp::client::AuthenticationSettings auth_settings;
+    auth_settings.provider = []() { return "invalid"; };
+
+    olp::client::OlpClientSettings client_settings;
+    client_settings.authentication_settings = auth_settings;
+    client_settings.task_scheduler = task_scheduler;
+    client_settings.network_request_handler = s_network;
+    client_settings.proxy_settings = GetLocalhostProxySettings();
+
+    return client_settings;
+  }
+
+ protected:
+  static std::shared_ptr<olp::http::Network> s_network;
+};
+
+std::shared_ptr<olp::http::Network> CatalogClientTest::s_network;
+
+namespace {
+
+void ClientThread(
+    const std::uint8_t client_id,
+    std::shared_ptr<olp::dataservice::read::CatalogClient> service_client,
+    std::string layer_id, std::chrono::milliseconds sleep_interval,
+    std::chrono::seconds runtime,
+    std::atomic<olp::http::RequestId>& request_counter) {
+  std::atomic_int success_responses{0};
+  std::atomic_int failed_responses{0};
+
+  std::mutex errors_mutex;
+  std::map<int, int> errors;
+
+  const auto end_timestamp = std::chrono::steady_clock::now() + runtime;
+
+  while (end_timestamp > std::chrono::steady_clock::now()) {
+    const auto partition_id = request_counter.fetch_add(1);
+
+    auto request = olp::dataservice::read::DataRequest()
+                       .WithLayerId(layer_id)
+                       .WithPartitionId(std::to_string(partition_id));
+
+    service_client->GetData(
+        request, [&](olp::dataservice::read::DataResponse response) {
+          if (response.IsSuccessful()) {
+            success_responses.fetch_add(1);
+          } else {
+            failed_responses.fetch_add(1);
+
+            // Collect errors information
+            std::lock_guard<std::mutex> lock(errors_mutex);
+            const auto& error = response.GetError();
+            const auto error_code = static_cast<int>(error.GetErrorCode());
+            const auto error_it = errors.find(error_code);
+            if (error_it != errors.end()) {
+              error_it->second++;
+            } else {
+              errors[error_code] = 0;
+            }
+          }
+        });
+
+    std::this_thread::sleep_for(sleep_interval);
```

Comment:
In case of sleep, we won't keep requested frequency of requests. Consider to calculate expected end time of iteration at the start of the iteration, and at the end of iteration we can wait remaining time.

Reply:
We don't need to be so precise.